### ISSUE TITLE
Failing multiple rules prints only the first error message

### ISSUE THREAD

**Issue by nh2** [ndmitchell/shake#254]
On shake-0.14.3, with these rules:

```haskell
    phony "fail1" $ do liftIO (putStrLn "print1" >> threadDelay 1000000 >> putStrLn "waited1" >> fail "fail1")
    phony "fail2" $ do liftIO (putStrLn "print2" >> threadDelay 2000000 >> putStrLn "waited2" >> fail "fail2")
    phony "fail3" $ do liftIO (putStrLn "print3" >> threadDelay 3000000 >> putStrLn "waited3" >> fail "fail3")
```

when running with command line arguments `fail1 fail2 fail3 --keep-going`, I get this output:

```
print3
print2
print1
waited1
Error when running Shake build system:
* fail1
user error (fail1)
Continuing due to staunch mode, this error will be repeated later
waited2
waited3
Error when running Shake build system:
* fail1
user error (fail1)
```

Shouldn't that somewhere include the messages that `fail "fail2"` and `fail "fail3"` should print? It seems they get lost.

**Comment by ndmitchell** [comment]
It took me half an hour to figure out why they don't all print...

The logic responsible for what you are seeing is that Shake doesn't print errors when they happen, but only when they escape to the top-level. In particular, a build system can have errors, yet still result in a successful build. If you used to depend on `foo.c`, but that file got deleted, you will often end up trying to rebuild with `foo.c`, have that fail, then rerun the rule that depended on it, and this time it won't depend on it so the build will succeed even though a rule fails. Building the children "speculatively" before they are needed is essential to cut out rebuilding of things that don't change, and thus errors aren't all fatal.

Because errors aren't necessarily harmful, I don't report them when happen (which is what you seem to be expecting), but only when an erroneous rule results in a problem, specifically a top-level action causing an exception. In your case, all three fail nodes have the same fake parent node. The first fails which causes the parent to fail, and then staunch mode flushes the queue before reporting the error. While the other two both failed, they didn't have anyone demand them, so Shake doesn't print anything.

As a fix for this example (but perhaps not the original problem you started from...) if I change `want` to be `mapM_ (action . need . return)` then I get three errors printed, as all the errors escape to the top-level.

Another alternative would be instead of having error nodes in the database store a single exception, to store a non-empty list. However, that's less efficient, as currently I continue as soon as I see one exception, while if it was a list, I'd have to check everything (building all the valid and invalid nodes).

So, in summary, changing `want` fixes it here - does it fix it in the original example you had as well? If so, I might be tempted to do that. Otherwise, I think it's just an inevitable consequence of the way Shake is structured.

**Comment by ndmitchell** [comment]
Btw, the `want` fix makes it work if you do a single top-level `want` for all three, or if you pass them on the command line, but as soon as you do `phony "fails" $ need ["fail1","fail2","fail3"]` it doesn't work again.

**Comment by nh2** [comment]
> If you used to depend on foo.c, but that file got deleted, you will often end up trying to rebuild with foo.c, have that fail, then rerun the rule that depended on it, and this time it won't depend on it so the build will succeed even though a rule fails. Building the children "speculatively" before they are needed is essential to cut out rebuilding of things that don't change, and thus errors aren't all fatal.

I don't fully understand this example yet - could you elaborate it a bit more? Are you describing a case where in the first run you dynamically discover a dependency, build it, and a second run where it's not a depenency any more, but since it's still in the database, Shake tries to rebuild it first?

> So, in summary, changing want fixes it here - does it fix it in the original example you had as well?

I *believe* that it won't fix it.

My current use case is a data processing setup where I have a phony target that discovers all the input data, and builds all the output data, like this:

```haskell
phony "build-all" $ do
  inFiles <- getDirectoryContents "inputs" ["//input.txt"]
  need [ "outputs" </> f </> "result.txt" | f <- inFiles ]

("outputs" <//> "result.txt") %> \out -> do
  let input = dropDirectory1 out </> "input.txt"
  () <- cmd "myprogram" [input]
  return ()
```

In this case, any `myprogram` invocation failing is a real failure that I can't recover from. I'm using `--keep-going` to build as many results as possible, but once that's done, I'd want to see all the error messages of the failed invocations so that I can investigate what's wrong with those inputs / with `myprogram`. Currently, I only see the first error propagate up, and the others are hidden.

**Comment by ndmitchell** [comment]
As an example that shows why an exception isn't always an error, consider the program:

    "foo" %> \_ -> copyFile "bar" "foo"
    "bar" %> \_ -> do src <- readFile' "baz"; writeFileChanged "bar" src

When building `foo`, I always build `bar` first, because if `bar` doesn't change then I can skip rebuilding `foo`. However, the user might delete `baz` and simultaneously change the rule for `foo` to:

    "foo" %> \_ -> writeFile' "foo" "x"

Since Shake can't know that the rule for `foo` changed, it still rebuilds `bar`, and that gives an error. It then rebuilds `foo`, decides it didn't really need `bar` after all, and thus ignores the error.

#### What to do next

To get what you need now, you could use `actionOnException` to detect failures in each of the translations and in the `IO` do a print. That would give you something right now.

Let's ignore the `want` change entirely. The fact that the two forms are different is confusing anyway, and might be a sign that something should be improved.

One option would be to always print all exceptions when they occur, even if they happen inside a rule that may never actually be required. That seems plausible. The only ugly bit is that normally you get one exception, then immediately get the same exception again. Therefore, perhaps it should only be turned on when you have staunch mode set? In addition, I'd refactor it so that instead of staunch mode bubbling the exception up the top before printing it, I can drop that. That's currently my favoured option - what do you think?

**Comment by nh2** [comment]
> Since Shake can't know that the rule for foo changed, it still rebuilds bar, and that gives an error.

Ah, so that's the main gap in my understanding - you actualy want to support the use case of the rules changing, but not rebuilding everything.

From the docs of `shakeVersion` ("The version number of your build rules. Change the version number to force a complete rebuild, such as when making significant changes to the rules that require a wipe") I always used the approach in https://github.com/ndmitchell/shake/issues/220, so I didn't even consider that what you described now was a supported feature.

> One option would be to always print all exceptions when they occur, even if they happen inside a rule that may never actually be required.

> In addition, I'd refactor it so that instead of staunch mode bubbling the exception up the top before printing it, I can drop that.

These two seem sensible to me. I think it's fair to print the exception if somebody does a significant change to the rules as you describe.

Intuitively I would argue that exceptions that are not caught explicitly by the user (here Shake catches them because it *may* recover from them) should be visible - they may be exactly what the user wants (in which case it's useful to see that the expected behaviour occurs), but they may also be something the user didn't expect (in which case it's even more useful to see them).

Independently, I believe that carefully changing build rules without bumping `shakeVersion` is its own craft, and would deserve its own section in the manual, explaining cases like the one you showed here. There we could also mention that doing that may incur exceptions that may be recovered from.
In general I'd appreciate a tips and tricks section about this topic - currently my approach is very conservative with the hash, and I'd probably like to avoid recompiling due to that in cases where I can.

**Comment by ndmitchell** [comment]
Done. Please give it a test, since anything with exceptions has a habit of being subtle!
